pgmpy/factors/Factor.py

```diff
@@ -496,11 +496,24 @@ def __eq__(self, other):
         if type(self) != type(other):
             return False
 
-        if self.variables == other.variables and all(self.cardinality == other.cardinality) \
-                and all(self.values == other.values):
-            return True
-        else:
+        if set(self.scope()) != set(other.scope()):
             return False
+        else:
+            order_dict = {}
+            for var_index in range(len(self.scope())):
+                order_dict[var_index] = other.scope().index(self.scope()[var_index])
+            if any(self.cardinality[[order_dict[i] for i in range(len(self.cardinality))]] != other.cardinality):
+                return False
+            indexes = np.array(list(itertools.product(*[range(i) for i in self.cardinality])))
```

Comment:
Okay. Here are my thoughts: 
* See the whole purpose of using ```numpy.array``` is not just indexing it is also about memory efficiency by having homogeneous data structure with fixed memory size. Creating a list would diminish the later advantage.
* Furthermore converting a generator into a list also looses the advantage of generator over the list.
* We can create this array with very basic implementation (without harming the code readability). We would be knowing the size of array in advance. So just create an array of zeros of that size (```np.zeros(...)```) and fill the elements as they are generated on the fly.
* I think (not pretty sure) that it would be efficient both in terms of memory and time. 

What are your thoughts on this?